Reduce JSON deserialization calls, add benchmarks

z4kn4fein opened the pull request:
### Describe the purpose of your pull request

Reducing the number of JSON deserialization calls by caching the last deserialized object in memory. Now, the deserialization is called only when the retrieved JSON string differs from the latest cached one.

Performance benchmark:
```
|      Method |       Mean |     Error |    StdDev | Ratio |  Gen 0 |  Gen 1 | Allocated |
|------------ |-----------:|----------:|----------:|------:|-------:|-------:|----------:|
| OldGetValue | 109.117 us | 0.2605 us | 0.2436 us |  1.00 | 2.9297 | 0.1221 |     25 KB |
| NewGetValue |   7.404 us | 0.0256 us | 0.0214 us |  0.07 | 0.9308 |      - |      8 KB |
```

### Related issues (only if applicable)

#22 

### Requirement checklist (only if applicable)

- [ ] I have covered the applied changes with automated tests.
- [x] I have executed the full automated test set against my changes.
- [x] I have validated my changes against all supported platform versions.

endret commented on src/ConfigCatClient/ConfigDeserializer.cs:
```diff
@@ -27,15 +28,19 @@ public bool TryDeserialize(ProjectConfig projectConfig, out IDictionary<string,
             }
+
+            using (var stringReader = new StringReader(config))
+            using (var reader = new JsonTextReader(stringReader))
```
curly brackets please

z4kn4fein replied:
{ d }
{ o }
{ n }
{ e }

endret commented on src/ConfigCatClient/ConfigDeserializer.cs:
```diff
@@ -9,16 +8,18 @@ internal class ConfigDeserializer : IConfigDeserializer
     {
         private readonly ILogger logger;
         private readonly JsonSerializer serializer;
+        private int hash;
```
I thinking about what happen when multiple threads try to write these fields: `hash` and `lastSerializedSettings`...
IMHO this solution would be safer if we can apply some locking (eg: ReaderWriterLockSlim)

z4kn4fein replied:
I think the worst thing that could happen is concurrent threads are hitting the hash check at the same time. That will degrade the performance as they will do the deserialization simultaneously, however, I think it's still better than acquiring upgradable read locks on every execution. 

I don't see logical drawbacks either as each thread will get its requested result.

I'm more afraid of hash collisions... but c'mon let's play dangerous :D

z4kn4fein commented:
### Update
Replaced `.GetHashCode()` with `MurmurHash3`. Far less chance for hashes to collide.

Updated benchmark:
```
| Method |       Mean |     Error |    StdDev | Ratio |  Gen 0 |  Gen 1 | Allocated |
|------- |-----------:|----------:|----------:|------:|-------:|-------:|----------:|
|    Old | 107.847 us | 0.4968 us | 0.4647 us |  1.00 | 2.9297 | 0.1221 |     25 KB |
|    New |   6.921 us | 0.0347 us | 0.0325 us |  0.06 | 1.3809 | 0.0076 |     11 KB |
```
